src/layouts/BaseLayout/hooks/useUserAccessGate.js

```diff
@@ -31,47 +34,88 @@ function useOnboardingRedirect({ username }) {
 
 const useUserAccessGate = () => {
   const { provider } = useParams()
-  const { termsOfServicePage, defaultOrgSelectorPage } = useFlags({
-    termsOfServicePage: false,
-    defaultOrgSelectorPage: false,
-  })
-  const { data, isLoading, isSuccess } = useUser({
+  const currentRoute = useRouteMatch()
+
+  const { termsOfServicePage, defaultOrgSelectorPage, sentryLoginProvider } =
+    useFlags({
+      termsOfServicePage: false,
+      defaultOrgSelectorPage: false,
+      sentryLoginProvider: false,
+    })
+
+  const {
+    data: userData,
+    isLoading: userIsLoading,
+    isSuccess: userIsSuccess,
+    isFetching: userIsFetching,
+  } = useUser({
     suspense: false,
-    enabled: !!provider || !config.IS_SELF_HOSTED,
+    enabled: !!provider && !config.IS_SELF_HOSTED,
   })
 
-  useOnboardingRedirect({ username: data?.user.username })
+  useOnboardingRedirect({ username: userData?.user?.username })
 
-  const isGuest = !data && isSuccess
-  let showAgreeToTerms = false,
-    showDefaultOrgSelector = false
+  const {
+    data: internalUser,
+    isLoading: internalUserIsLoading,
+    isSuccess: internalUserIsSuccess,
+    isFetching: internalUserIsFetching,
+  } = useInternalUser({
+    suspense: false,
+    enabled: !config.IS_SELF_HOSTED,
+  })
+
+  const foundUser = !userData && userIsSuccess
+  const foundInternalUser = !internalUser && internalUserIsSuccess
+  const isGuest = foundUser || foundInternalUser
+
+  let showAgreeToTerms = false
+  let showDefaultOrgSelector = false
+  let redirectToSyncPage = false
+
+  // the undefined provider check can be removed when the ToS has
+  // been refactored to no longer use a provider
+  if (
+    termsOfServicePage &&
+    !isUndefined(provider) &&
+    !isGuest &&
+    !config.IS_SELF_HOSTED
+  ) {
+    showAgreeToTerms = userData?.termsAgreement === false
+  }
 
-  // If we don't have a provider, we cant check the TOS agreement, so we alow the user to access the full experience.
-  if (!provider) {
-    return {
-      isFullExperience: true,
-      isLoading: false,
-      showAgreeToTerms,
-      showDefaultOrgSelector,
-    }
+  const onSyncPage = currentRoute.path === '/sync'
```

Comment:
can you do `const match = useRouteMatch('/sync');` to check if match.isExact, it directly checks if there is an exact match for the '/sync' route, which is more concise and clearer in its intent

Reply:
The reason why I have done it like this is because it is expandable in the future and not limited just to the `/sync` route. I think it is better left this way when/if the time comes we can quickly check the path without having any hassle.